server/etcdserver/v3_server.go

```diff
@@ -279,6 +289,12 @@ func (s *EtcdServer) LeaseRevoke(ctx context.Context, r *pb.LeaseRevokeRequest)
 }
 
 func (s *EtcdServer) LeaseRenew(ctx context.Context, id lease.LeaseID) (int64, error) {
+	if s.isLeader() {
```

Comment:
I haven't dug too dip into analyzing correctness of this solution, however when comparing to [V2 solution](https://github.com/etcd-io/etcd/pull/7015) I found strange that we treat members differently. In your solution you only wait for applied commits on leader, however original solution [waits on any member ](https://github.com/fanminshi/etcd/blob/2faf72f47c2ce1633ed93035473fb7a9ecc771a4/lease/leasehttp/http.go#L69-L74). Are we sure that by checking this only on leader we are not missing any edge case (for example race with leader election)? 

Please let me know if I misunderstood anything as I'm not supper familiar with this part of the codebase.

Reply:
Generally speaking, there are two possibilities:
1. Previous it's leader, and it changes to non-leader;
2. Previous it's not a leader, and it changes to a leader;

The first case happens in [server.go#L760](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/server.go#L760). In this case, the `s.lessor.Renew` will return a `ErrNotPrimary` error, then the etcd member will forward the request to the new leader. It should be OK in this case.

The second case happens in [server.go#L1877](https://github.com/etcd-io/etcd/blob/main/server/etcdserver/server.go#L1877). It is in the apply path, which means it can only happen in a separate apply workflow. In current apply workflow, it can never happen based on current implementation. 

So we should be good in either case.

Reply:
Ok, could we make it consistent between V3 and V2 lease code?

Reply:
Actually no need to make it consistent, a couple of reasons: 
1. The existing logic in [http.go](https://github.com/etcd-io/etcd/blob/main/server/lease/leasehttp/http.go#L49) is only for peer communication.  Only the leader can receive the lease http requests, so no need to check whether it's the leader any more. 
2. Actually from implementation perspective, the [leaseHandler](https://github.com/etcd-io/etcd/blob/main/server/lease/leasehttp/http.go#L44) doesn't have a reference to the Etcdserver, so it doesn't know whether it's the leader;
3. Even in the case of leader changing, there is also a check inside the [lessor](https://github.com/etcd-io/etcd/blob/main/server/lease/lessor.go#L393).